src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -199,46 +201,45 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
         }
     }
 
-    /// Add the given qualification to self.qualif.
+    /// Add the given qualification to `self.qualif`.
     fn add(&mut self, qualif: Qualif) {
-        self.qualif = self.qualif | qualif;
+        self.qualif.mut_interior |= qualif.mut_interior;
+        self.qualif.needs_drop |= qualif.needs_drop;
+        self.qualif.not_const |= qualif.not_const;
     }
 
-    /// Add the given type's qualification to self.qualif.
+    /// Add the given type's qualification to `self.qualif`.
     fn add_type(&mut self, ty: Ty<'tcx>) {
-        self.add(Qualif::MUTABLE_INTERIOR | Qualif::NEEDS_DROP);
-        self.qualif.restrict(ty, self.tcx, self.param_env);
+        self.qualif.mut_interior = !ty.is_freeze(self.tcx, self.param_env, DUMMY_SP);
+        self.qualif.needs_drop = ty.needs_drop(self.tcx, self.param_env);
     }
 
-    /// Within the provided closure, self.qualif will start
+    /// Within the provided closure, `self.qualif` will start
     /// out empty, and its value after the closure returns will
     /// be combined with the value before the call to nest.
     fn nest<F: FnOnce(&mut Self)>(&mut self, f: F) {
         let original = self.qualif;
-        self.qualif = Qualif::empty();
+        self.qualif = Qualif::default();
         f(self);
         self.add(original);
     }
 
     /// Assign the current qualification to the given destination.
     fn assign(&mut self, dest: &Place<'tcx>, location: Location) {
         trace!("assign: {:?}", dest);
-        let qualif = self.qualif;
-        let span = self.span;
-        let store = |slot: &mut Option<Qualif>| {
-            if slot.is_some() {
-                span_bug!(span, "multiple assignments to {:?}", dest);
-            }
-            *slot = Some(qualif);
+        let store = |this: &mut Self, index| {
+            this.local_mut_interior.set_member(&index, this.qualif.mut_interior);
+            this.local_needs_drop.set_member(&index, this.qualif.needs_drop);
+            this.local_not_const.set_member(&index, this.qualif.not_const);
         };
 
         // Only handle promotable temps in non-const functions.
         if self.mode == Mode::Fn {
             if let Place::Local(index) = *dest {
                 if self.mir.local_kind(index) == LocalKind::Temp
```

Comment:
I believe the failure [here](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/issue-49955.rs#L19) is due to this not also entering this arm in case of `LocalKind::ReturnPointer`. Although I don't see how that worked before!?